src/python/WMComponent/WorkflowUpdater/WorkflowUpdaterPoller.py

```diff
@@ -101,13 +260,86 @@ def algorithm(self, parameters=None):
-
+            self.adjustJSONSpec(puWflows, msPileupList)
         except Exception as ex:
             msg = f"Caught unexpected exception in WorkflowUpdater. Details:\n{str(ex)}"
             logging.exception(msg)
             raise WorkflowUpdaterException(msg) from None
 
+    def adjustJSONSpec(self, puWflows, msPileupList, dest=None):
+        """
+        Main logic of the algorithm:
+        - for every pileup record find out location of tarball
+        - get configuration files within tarball
+        - extract block information
+        - replace pileupconf.json files within tarball
+        :param puWflows: list of active pileup workflows with the following structure:
+            {"name": string with workflow name,
+             "spec": string with spec path,
+             "pileup": list of strings with pileup names}
+        :param msPileupList: list of all pileup records in MSPileup service. It has the following structure:
+            {"pileupName": string with pileup name,
+             "customName": string with custom pileup name - if any,
+             "rses": list of RSE names,
+             "blocks": list of block names}
+        :param dest: optional destination parameter to write final tar ball (use for unit tests)
+        :return: nothing, it performs checks and adjust in place pileupconf.json file(s)
+        """
+        # loop over active pileup workflows
+        for wflow in puWflows:
+            # this logic implies that we have untarred workflow tar ball
+            sdir = os.path.join(self.sandboxDir, wflow['name'])
+            tarName = wflow['name'] + '-Sandbox.tar.bz2'
+            tarFile = os.path.join(sdir, tarName)
+            self.logger.debug("process %s, tarfile %s", wflow, tarFile)
+
+            # json dictionary to write back to our tarball
+            jdict = {}
+
+            # list over pileup workflows in MSPileup service
+            for pileupDoc in msPileupList:
+                # check if active pileup workflow is found in MSPileup one
+                pileupName = pileupDoc['pileupName']
+                if pileupName in wflow["pileup"]:
+                    # find pileup configuration files in sandbox area
+                    self.logger.debug("find pileupconf.json in %s", sdir)
+                    files = findFiles(sdir, "pileupconf.json")
+                    self.logger.debug("found matched pileup %s with conf files %s", pileupName, files)
+
+                    # check config files content if they have the same or different datasets
+                    for fname in files:
+                        # make sure that there is no double slaehes in a file path
+                        fname = fname.replace("//", "/")
+                        self.logger.debug("search %s in %s", pileupName, fname)
+                        pileupConfContent = ""
+                        with open(fname, 'r', encoding='utf-8') as istream:
```

Comment:
I think we agreed that first we should get the logic right and then we can look into optimization. Yes, it will be done but I do not want waste time of moving parts around until I have correct logic in place.